### PR TITLE
test: user tests [INFENG-455]

### PR DESCRIPTION
**Pull request by JComins000**
## Ticket

[INFENG-455]

## Description

user management page tests

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/e2e/tests/userManagement.spec.ts [hunk 1]
```diff
@@ -19,6 +19,35 @@ test.describe('User Management', () => {
...
+      pagination.perPage.perPage10,
+      pagination.perPage.perPage20,
+      pagination.perPage.perPage50,
+      pagination.perPage.perPage100,
+    ]) {
+      await pagination.perPage.pwLocator.click();
+      await paginationOption.pwLocator.click();
+      await expect(userManagementPage.skeletonTable.pwLocator).not.toBeVisible();
+      const matches = (await pagination.perPage.pwLocator.innerText()).match(/(\d+) \/ page/);
+      if (matches === null) {
```
Comment (djanicekpach) [inline webui/react/src/e2e/tests/userManagement.spec.ts:41]: Can we use an expect() or assertion library for this? It seems like a test assertion more than a runtime error.

File: webui/react/src/e2e/tests/userManagement.spec.ts [hunk 2]
```diff
@@ -19,6 +19,35 @@ test.describe('User Management', () => {
...
+      throw new Error('Number not present in tab.');
+    }
+    const expetedRowCount = +match[0];
+    for (const paginationOption of [
+      pagination.perPage.perPage10,
+      pagination.perPage.perPage20,
+      pagination.perPage.perPage50,
+      pagination.perPage.perPage100,
+    ]) {
+      await pagination.perPage.pwLocator.click();
```
Comment (djanicekpach) [inline webui/react/src/e2e/tests/userManagement.spec.ts:37]: could we make each of these a test.step() so we can see on failure which one failed quickly? IDK if playwright lets us put steps and for loops or not, but could be nice to have.

**Review by djanicekpach — APPROVED** [review]
looks good, a couple of nice-to-have comments.

File: webui/react/src/e2e/tests/userManagement.spec.ts (hunk 1 above)
Comment (JComins000) [reply webui/react/src/e2e/tests/userManagement.spec.ts:41]: image

File: webui/react/src/e2e/tests/userManagement.spec.ts [hunk 3]
```diff
@@ -19,48 +19,95 @@ test.describe('User Management', () => {
...
-    let userid: string;
-    let username: string;
+    let testUser: User;
 
     test.beforeAll(async ({ browser }) => {
-      username = 'test-user-' + uuidv4();
-
       await test.step('Login', async () => {
-        page = await browser.newPage();
+        page = await browser.newPage({ recordVideo: { dir: './src/e2e/test-results' } });
```
Comment (djanicekpach) [inline webui/react/src/e2e/tests/userManagement.spec.ts:70]: Is this a debug thing? We should be setting up video recording in the config. I'd prefer not to have it setup on a test-by-test basis

**Review by djanicekpach — APPROVED** [review]
One question but otherwise looks good

File: webui/react/src/e2e/tests/userManagement.spec.ts (hunk 3 above)
Comment (JComins000) [reply webui/react/src/e2e/tests/userManagement.spec.ts:70]: this is required to get video on the "second page". If I wasnt doing this `new Page` then I wouldn't need to save this video. only problem is that it doesn't show in the same folder as the test videos

**Review by dannysauer — APPROVED** [review]

File: webui/react/playwright.config.ts [hunk 4]
```diff
@@ -6,6 +6,18 @@ import { defineConfig, devices } from '@playwright/test';
...
+const server_addess = process.env.PW_SERVER_ADDRESS
+let port: number;
+if (server_addess === undefined) {
+  port = 3001
+} else {
+  const match = server_addess.match(/\d+/)
+  if (match === null || isNaN(+match)) {
+    throw new Error(`Expected port number in PW_SERVER_ADDRESS ${server_addess}, ${match}`)
+  }
+  port = +match
```
Comment (keita-determined) [inline webui/react/playwright.config.ts:18]: `match` return [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match#return_value), so if i dont miss anything, `port = +match` wouldn't work as expected. Anyways i think `Number(server_addess)` is preferable

Comment (JComins000) [reply webui/react/playwright.config.ts:18]: i think + worked as a shortcut. I wonder what it does under the hood. Anyway, im just about to push a commit that fixes it

File: webui/react/playwright.config.ts [hunk 5]
```diff
@@ -6,6 +6,14 @@ import { defineConfig, devices } from '@playwright/test';
 import * as dotenv from 'dotenv';
 dotenv.config();
 
+let port: number;
+const serverAddess = process.env.PW_SERVER_ADDRESS
+if (serverAddess === undefined) {
+  throw new Error('Expected PW_SERVER_ADDRESS to be set.')
+} else {
+  port = Number(new URL(serverAddess).port)
```
Comment (keita-determined) [inline webui/react/playwright.config.ts:14]: this looks good, but we might want to check if `port` exist. `Number()` returns `NaN` if input isnt number. For example, `https://localhost` doesnt work. also, i dont think else clause is needed. we can directly assign value with `const` instead of `let`

File: webui/react/playwright.config.ts [hunk 6]
```diff
@@ -6,6 +6,15 @@ import { defineConfig, devices } from '@playwright/test';
 import * as dotenv from 'dotenv';
 dotenv.config();
 
+const serverAddess = process.env.PW_SERVER_ADDRESS
+if (serverAddess === undefined) {
+  throw new Error('Expected PW_SERVER_ADDRESS to be set.')
+}
+const port = Number(new URL(serverAddess).port)
```
Comment (keita-determined) [inline webui/react/playwright.config.ts:13]: one more edge case,  
idk if its clean but 

```suggestion
const port = Number(new URL(serverAddess).port || 'not exist')
```

**Review by keita-determined — APPROVED** [review]

File: webui/react/src/e2e/tests/userManagement.spec.ts (hunk 3 above)
Comment (JComins000) [reply webui/react/src/e2e/tests/userManagement.spec.ts]: I ended up changing the tests to use the "other page" and so now I don't need to save a video from this one anymore. changes were made [here](https://github.com/determined-ai/determined/compare/jcom/INFENG-455/user-tests-continued?expand=1). i figure we can review this one separately, merge it into this PR, and merge it all into main at once

File: webui/react/src/e2e/tests/userManagement.spec.ts [hunk 7]
```diff
@@ -19,48 +20,117 @@ test.describe('User Management', () => {
     await expect(page).toHaveURL(userManagementPage.url);
   });
 
+  test.skip('Users table count matches admin page users tab', async ({ page }) => {
```
Comment (JComins000) [inline webui/react/src/e2e/tests/userManagement.spec.ts:23]: skip until this PR
https://github.com/determined-ai/determined/pull/9214/commits

File: webui/react/playwright.config.ts [hunk 8]
```diff
@@ -23,7 +23,7 @@ export default defineConfig({
   fullyParallel: !!process.env.CI,
 
   /* https://playwright.dev/docs/test-timeouts#global-timeout */
-  globalTimeout: process.env.PWDEBUG ? 0 : 5 * 60 * 1000, // 3 min unless debugging
+  globalTimeout: process.env.PWDEBUG ? 0 : 500_000, // 500s unless debugging
```
Comment (djanicekpach) [inline webui/react/playwright.config.ts:26]: can we revisit this in the next PR now that devcluster is in and there aren't 1500 users?

Comment (JComins000) [reply webui/react/playwright.config.ts:26]: we also have another ticket to simply revisit all timeouts later. i'll remove the change for now

File: webui/react/src/e2e/models/pages/Admin/UserManagement.ts [hunk 9]
```diff
@@ -79,7 +79,13 @@ export class UserManagement extends AdminPage {
     });
     expect(
       filteredRows,
-      `name:${name}, component:${await this.table.table.rows.pwLocator.allInnerTexts()}`,
+      `name: ${name}
+      users: ${(await this.table.table.rows.user.pwLocator.allInnerTexts()).map((item) =>
+        item.replace(/(\r\n|\n|\r)/gm, ''),
```
Comment (djanicekpach) [inline webui/react/src/e2e/models/pages/Admin/UserManagement.ts:84]: nitpick: can we pull this map+regex out into it's own constant/function with a descriptive name? Or is there a built in function for `trim`ing whitespace?

**Review by djanicekpach — APPROVED** [review]
